lint ruby drc code.

atorkmabrains opened the pull request:
Fixes #14 

Please don't merge this PR before the other PRs.

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

proppy commented:
can you invert the order of the commits in the history in order to:
- only keep in this PR: the changes that add lint to ci w/ dualgate modification to pass the checks
- rebase the other PRs to include the corresponding lint modification for each deck check

proppy commented:
That will make it easier to review things concurrently.

atorkmabrains commented:
@FaragElsayed2 Could you please do that? My bad.

atorkmabrains commented:
@proppy If you accept all the other open PRs and then this one in that order, everything will fall in place as you want. The rest of the tables will be correct moving forward and the new tables will be lint clean.

We won't be able to work on this until Sunday. If you can do the above that would be really helpful to speed up the process on our end.

FaragElsayed2 commented:
@proppy I have removed all other files as requested and only kept dualgate drc file and linting checks for ruby files.

FaragElsayed2 commented:
@proppy And @mithro Please add me as a collaborator for this project.

proppy commented:
> @proppy And @mithro Please add me as a collaborator for this project.

please ask mithro@ directly (let's keep the comment in this issue about the actual review)

proppy commented:
> @proppy I have removed all other files as requested and only kept dualgate drc file and linting checks for ruby files.

Thanks! that makes it a lot easier to review. Can you rebase 531323a on top of the main branch to clean up the history?

proppy commented:
> @proppy If you accept all the other open PRs and then this one in that order, everything will fall in place as you want. The rest of the tables will be correct moving forward and the new tables will be lint clean.
> We won't be able to work on this until Sunday. If you can do the above that would be really helpful to speed up the process on our end.

I prefer to review the PR as it is currently as it will allow us to review each further DRC PR idependently with the lint workflow enabled.

proppy commented on Makefile:
```diff
@@ -25,12 +25,22 @@ ENVIRONMENT_FILE := pdk_regression.yml
 	@$(IN_CONDA_ENV) flake8 .
 
+# Lint ruby code
+lint_ruby:| $(CONDA_ENV_PYTHON)
```
should we add `ruby` as a dependencies in the conda env?

FaragElsayed2 replied:
Same as below

proppy commented on Makefile:
```diff
@@ -25,12 +25,22 @@ ENVIRONMENT_FILE := pdk_regression.yml
 
+# Lint ruby code
+lint_ruby:| $(CONDA_ENV_PYTHON)
+	@$(IN_CONDA_ENV) gem install rubocop
```
Can we add this in a Gemfile and install the dependencies as part of the conda-env bootstrap?

FaragElsayed2 replied:
Referring to #38

proppy commented on Makefile:
```diff
@@ -25,12 +25,22 @@ ENVIRONMENT_FILE := pdk_regression.yml
+	@$(IN_CONDA_ENV) gem install rubocop
+	@$(IN_CONDA_ENV) rubocop=`which rubocop`
+	@$(IN_CONDA_ENV) rubocop_dir=`dirname $$rubocop`
+	@$(IN_CONDA_ENV) echo $$rubocop_dir
```
do you want to leave this?

FaragElsayed2 replied:
Same as below

proppy replied:
I was referring to the `echo` statement, which seems to be there for troubleshooting?

FaragElsayed2 replied:
Removed

proppy commented on Makefile:
```diff
@@ -25,12 +25,22 @@ ENVIRONMENT_FILE := pdk_regression.yml
+	@$(IN_CONDA_ENV) rubocop=`which rubocop`
+	@$(IN_CONDA_ENV) rubocop_dir=`dirname $$rubocop`
+	@$(IN_CONDA_ENV) echo $$rubocop_dir
+	@$(IN_CONDA_ENV) ln -s ../../../bin/ruby $$rubocop_dir/ruby
```
that shouldn't be necessary if you install ruby in the env.

FaragElsayed2 replied:
I have added ruby as a dependency in the `pdk_regression.yml` file and it still causes an issue.

![image](https://user-images.githubusercontent.com/91015308/216909864-650ef6fa-4cb7-4d72-872a-703851d0d57d.png)

proppy replied:
by default `ruby` shows up `$CONDA_PREFIX/bin` for me when installing it with conda, can you clarify:
- which version of ruby got installed
- which version of rubygem got installer
- how you are installing rubocop

maybe this could be an acceptable workaround:
```
ln -s $CONDA_PREFIX/bin/ruby $CONDA_PREFIX/share/rubygems/bin/ruby
```
wdyt?

FaragElsayed2 replied:
I have added ruby in yml file as shown:

```yaml
name: pdk_regression
channels:
  - litex-hub
  - defaults
dependencies:
  - klayout
  - python
  - ruby
  - pip
  - pip:
    - -r requirements.txt
```
I have installed rubocop using `gem install rubocop`

The solution you have proposed above is exactly the same as the one existed before.

Please advise how to move forward.

FaragElsayed2 replied:
@proppy 

Please see below the versions of ruby and gem in the conda environment 

```
Ruby: ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x86_64-linux]
Gem: 3.3.7
```

proppy replied:
I'd prefer for is to use $CONDA_PREFIX to create the link and to track the ruby dependencies using `Gemfile`

proppy commented on .rubocop.yml:
```diff
@@ -0,0 +1,45 @@
+# where the inspected file is and continue its way up to the root directory.
+#
+# See https://docs.rubocop.org/rubocop/configuration
+
```
can you  add a comment about why we're disabling those? (most of them seems useful)

FaragElsayed2 replied:
We disabled the rules that generates a lot of linting issues to reduce the linting noise and we plan to clean up the code later as requested in #37

proppy replied:
#37 sounds unrelated to lintint, are you referring to #14?

FaragElsayed2 replied:
Yes, referring to #14

proppy replied:
This PR is marked as fixing #14, so unless we keep #14 open after merging it we should make sure we have another issue to follow up with more lintint changes.

I would recommend enabling those checks and addressing them here for the dualgate rules

FaragElsayed2 replied:
@proppy Let's move forward with the merge. I have opened new issue #40 to keep track of this required change.

proppy commented on klayout/drc/rule_decks/tail.drc:
```diff
@@ -14,10 +14,6 @@
-logger.info("%s DRC Total Run time %f seconds" % [table_name, run_time])
-
-
+logger.info(format('%<table>s DRC Total Run time %<time>f seconds', table: table_name, time: run_time))
```
any reason you don't use ruby string interpolation here?
see https://docs.ruby-lang.org/en/3.0/syntax/literals_rdoc.html#label-Strings

FaragElsayed2 replied:
Done

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 #================================================
-logger.info("Starting running GF180MCU Klayout DRC runset on %s" % [$input])
-logger.info("Ruby Version for klayout: %s" % [RUBY_VERSION])
+logger.info(format('Starting running GF180MCU Klayout DRC runset on %s', $input))
```
use string interpolation?

FaragElsayed2 replied:
Done

proppy replied:
I don't see any update, did you push?

FaragElsayed2 replied:
Not yet, I will push updates now.

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-logger.info("Starting running GF180MCU Klayout DRC runset on %s" % [$input])
-logger.info("Ruby Version for klayout: %s" % [RUBY_VERSION])
+logger.info(format('Starting running GF180MCU Klayout DRC runset on %s', $input))
+logger.info(format('Ruby Version for klayout: %s', RUBY_VERSION))
```
use string interpolation?

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
+  end
 end
 
+table_name = $table_name || 'main'
```
curious why you don't use the existing global variable here?
ex: `$table_name || = 'main'`

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-else
-  WEDGE = "true"
-end # WEDGE
+WEDGE = if $wedge == 'false'
```
do we want to default to be true here?

FaragElsayed2 replied:
Yes

proppy replied:
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
Original structure has some linting issues.
![image](https://user-images.githubusercontent.com/91015308/217463997-5521b747-8d79-45e0-9590-924c7200fa6e.png)

proppy replied:
Ah I see, are you intentionally keeping this a `string` rather that converting it to a boolean? I would assume that the rest of the rules would prefer to test this directly rather than doing string comparison?

proppy replied:
Can you comment here ?

FaragElsayed2 replied:
Done, bool check function is used for string to boolean conversion.

![image](https://user-images.githubusercontent.com/91015308/218083493-758fb593-2223-4eff-98fa-51a5ef1a36be.png)

FaragElsayed2 replied:
![image](https://user-images.githubusercontent.com/91015308/218083596-dfa2e576-40dd-492b-9230-e4ab055950b8.png)

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 
 # connectivity rules
-if $conn_drc == "true"
+if $conn_drc == 'true'
```
`$conn_dir ||= false`?

FaragElsayed2 replied:
Same as below, referring to #37

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 if $report
-    logger.info("GF180MCU Klayout DRC runset output at: %s" % [$report])
-    report("DRC Run Report at", $report)
+  logger.info(format('GF180MCU Klayout DRC runset output at: %s', $report))
```
string interpolation?

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-else
-  BALL = "true"
-end # BALL
+BALL = if $ball == 'false'
```
do we want to default to be true here?

FaragElsayed2 replied:
Yes

proppy replied:
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
Same as above

proppy replied:
Are you intentionally keeping this a `string` rather that converting it to a boolean? I would assume that the rest of the rules would prefer to test this directly rather than doing string comparison?

FaragElsayed2 replied:
Done, `bool check` function is used for string to boolean conversion.
![image](https://user-images.githubusercontent.com/91015308/218083375-9cfd64cd-e228-4e99-bb45-3ac53b862fff.png)

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-else
-  GOLD = "true"
-end # GOLD
+GOLD = if $gold == 'false'
```
do we want to default to be true here?

FaragElsayed2 replied:
Yes

proppy replied:
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
Same as above

proppy replied:
Ah I see, are you intentionally keeping this a `string` rather that converting it to a boolean? I would assume that the rest of the rules would prefer to test this directly rather than doing string comparison?

FaragElsayed2 replied:
Done, `bool check` function is used for string to boolean conversion.

![image](https://user-images.githubusercontent.com/91015308/218083010-48115e1e-46fa-4046-a3bd-965c3facd243.png)

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-else
-  MIM_OPTION = "B"
-end
+MIM_OPTION = $mim_option || 'B'
```
`$mim_option ||= 'B'`

FaragElsayed2 replied:
`MIM_OPTION` is a different variable than `mim_option`.

proppy replied:
I didn't understand why they need to be different but I think we can continue this conversion in #37 and focus on syntax fixes in this PR.

proppy replied:
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
Same as above

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-else
-  OFFGRID = true
-end # OFFGRID
+OFFGRID = $offgrid != 'false'
```
do we want to default to be true here?

FaragElsayed2 replied:
Yes

proppy replied:
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
Same as above

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 else
-    thr ||=  Etc.nprocessors
-    threads(thr)
+  thr ||= Etc.nprocessors
```
I think you want to use `$thr` here otherwise it won't consider the existing value.

FaragElsayed2 replied:
The if condition in line `111` checks if the variable is already defined or not.

proppy replied:
Yes bur this seems redundant with the usage of the `||=` operator.

Removing the if and doing:
```
$thr ||= Etc.nprocessors
threads($thr)
```
should be equivalent?

FaragElsayed2 replied:
Yes

Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 logger.info("Verbose mode: #{$verbose}")
-if $verbose == "true"
+if $verbose == 'true'
   verbose(true)
```
`verbose($verbose.to_b)` ?

FaragElsayed2 replied:
Same as below, referring to #37

proppy replied:
Agreed let's keep refactoring for #37

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
+METAL_TOP = $metal_top || '9K'
 
-logger.info("METAL_TOP Selected is %s" % [METAL_TOP])
+logger.info(format('METAL_TOP Selected is %s', METAL_TOP))
```
string interpolation?

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-else
-  METAL_TOP = "9K"
-end # METAL_TOP
+METAL_TOP = $metal_top || '9K'
```
any reason we don't keep using `$metal_top` here?

FaragElsayed2 replied:
`METAL_TOP` variable is different than `metal_top`. It's used in other tables.

FaragElsayed2 replied:
Referring to #37

proppy replied:
Agreed let's keep any config refactoring for #37?

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
+METAL_LEVEL = $metal_level || '5LM'
 
-logger.info("METAL_STACK Selected is %s" % [METAL_LEVEL])
+logger.info(format('METAL_STACK Selected is %s', METAL_LEVEL))
```
string interpolation?

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-  FEOL = "true"
-  logger.info("FEOL is enabled.")
-end # FEOL
+  FEOL = 'true'.freeze
```
curious why we call freeze here?

FaragElsayed2 replied:
Removed

FaragElsayed2 replied:
![image](https://user-images.githubusercontent.com/91015308/216932961-d4b63a94-420a-4f7a-b065-2d1c1553f2b2.png)

Linting issue

proppy replied:
ah make sense, I didn't realize that `FEOL` was a constant.

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-  BEOL = "true"
-  logger.info("BEOL is enabled.")
-end # BEOL
+  BEOL = 'true'.freeze
```
same q here?

FaragElsayed2 replied:
Removed

FaragElsayed2 replied:
Same as above

proppy replied:
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
`BEOL` switch is similar to `FEOL`.

![image](https://user-images.githubusercontent.com/91015308/217466056-077a4803-7f26-4bf7-b867-103f70831fe3.png)

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 
 # FEOL
-if $feol == "false"
+if $feol == 'false'
```
do we want to default to true here?

FaragElsayed2 replied:
Yes

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
 
 # BEOL
-if $beol == "false"
+if $beol == 'false'
```
do we want to default to true here?

FaragElsayed2 replied:
Yes

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,766 @@
-polygons_count  += count
+comp = get_polygons(22, 0)
+count = comp.count
+logger.info(format('comp has %d polygons', count))
```
use string interpolation here and below?

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -1019,17 +904,16 @@ def conn_space(layer,conn_val,not_conn_val, mode)
-  if conn_val > not_conn_val
-    raise "ERROR : Wrong connectivity implementation"
-  end
+def conn_separation(layer1, layer2, conn_val, not_conn_val, mode)
```
is this function used?

atorkmabrains replied:
Yes, in other tables. 'main' is for all tables.

proppy replied:
can we add it once it's actually used, otherwise it won't be covered by any tests.

atorkmabrains replied:
@proppy I prefer not to change `main` in every PR to make the PRs easier to merge. That's why I kept it here. And it's covered in the tables that uses this function.

proppy replied:
sure, as long as it's covered in the other table I guess it's fine!

proppy commented on .github/workflows/linting.yml:
```diff
@@ -29,4 +29,16 @@ jobs:
+  lint_ruby:
+    runs-on: ubuntu-latest
+    strategy:
+      max-parallel: 12
```
curious why we limit this?

FaragElsayed2 replied:
We don't have a specific reason but we just show 12 as a reasonable number. If you think we need to change that. Please let me know and propose a number.

proppy replied:
Is the default not good enough?

FaragElsayed2 replied:
It's good.

Should I remove it ?

proppy replied:
Yes, if the default is good enough no reason to override it :)

FaragElsayed2 replied:
Done

proppy reviewed (changes requested):
I'd recommend considering something like #37 to simplify the handling of configuration defaults .

atorkmabrains commented:
@proppy Could you please resolve any open discussions?

atorkmabrains commented:
@proppy Could you please merge all the changes as @FaragElsayed2 has already either addressed all your questions or opened an issue with it?

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,763 @@
-    end
-end
-
-if $table_name
```
any reaosn we got rid of the table_name check here?

FaragElsayed2 replied:
Added directly to `tail.drc` file where it is used

proppy replied:
I don't see it in `tail.drc`, did you push?

FaragElsayed2 replied:
Yes
![image](https://user-images.githubusercontent.com/91015308/217475446-6a168f1e-717e-4790-b3e4-5a99a83c33af.png)

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,763 @@
-end
-logger.info("Number of threads to use %s" % [$thr])
+# threads
+$thr ||= Etc.nprocessors
```
Can we keep the original structure of the code then and keep any config refactoring for #37?

FaragElsayed2 replied:
Done, referring to #37

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,763 @@
-  return $run_mode == "deep" ? ps : ps.merged
+def get_polygons(layer, data_type)
+  ps = polygons(layer, data_type)
+  $run_mode == 'deep' ? ps : ps.merged
```
shouldn't we keep the return here?

FaragElsayed2 replied:
No need, It works well without it.
![image](https://user-images.githubusercontent.com/91015308/217468535-567357ae-a135-4150-99c9-de9297ec8fdc.png)

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,763 @@
+      logger.info("via4 has #{count} polygons")
+      polygons_count += count
+
+      case METAL_LEVEL
```
seems weird to switch from if/else to a case/when in the middle of block?

FaragElsayed2 replied:
To fix this linting issue
![image](https://user-images.githubusercontent.com/91015308/217474256-1545df10-b67b-4bbf-bdb1-815220f1dc14.png)

proppy replied:
I'd recommend that you convert the whole block in a big case statement, since most of the processing depends of the numbers of layers

FaragElsayed2 replied:
@proppy I have added a new issue for that. Referring to #42

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,763 @@
-        count   = metal5_drawn.count()
-        logger.info("metal5_drawn has %d polygons" % [count])
-        polygons_count  += count
+      when '6LM'
```
seems weird to switch from if/else to a case/when in the middle of block?

FaragElsayed2 replied:
Same as above

FaragElsayed2 replied:
@proppy Referring to #42

proppy commented:
> @proppy Could you please merge all the changes as @FaragElsayed2 has already either addressed all your questions or opened an issue with it?

Resolved most of the comments (still a few left).

Can we make sure to squash the commits (since there is a lot of unrelated changes in the history)?

That should also make it easier to rebase the other PR on top of it.

atorkmabrains commented:
@FaragElsayed2 Squashing commits on github to avoid too many commits for the PR:
https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/squashing-commits

atorkmabrains commented:
@proppy I think @FaragElsayed2 has already addressed most of your concerns. Could you please merge the PR and open issues with any remaining issues as we agreed?

proppy commented:
@atorkmabrains I agree that most of the comment have been addressed, but as suggested before I'd like the history to be squashed before merging (there is currently an history of 106 commits which are mostly unrelated to this change)

atorkmabrains commented:
@FaragElsayed2 Could you please squish the commits history?

atorkmabrains commented:
@proppy Commits are squashed to one commit as requested.

atorkmabrains commented:
@proppy Could you please let me know if there is anything else?

proppy commented:
It looks good, but please use a more descriptive commit message (no need to mention it's in one commit)

proppy reviewed (approved):
Just some nits left above better handing of the conditions per layers. But feel free to address them in a separate change.

Thanks for the hard work on this.

FaragElsayed2 commented:
@proppy I think we have done with all issues, Could you please merge ?

proppy commented:
Sure can you make sure to address it the last remaining comment and reword the commit message?

Thanks in advance

FaragElsayed2 commented:
@proppy Done

atorkmabrains commented:
Thanks @FaragElsayed2

atorkmabrains commented:
@proppy Anything else?

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,772 @@
 # connectivity rules
-if $conn_drc == "true"
+if $conn_drc == 'true'
   CONNECTIVITY_RULES = $conn_drc
```
shouldn't you use `bool_check` here too?

FaragElsayed2 replied:
Done

proppy commented on klayout/drc/rule_decks/main.drc:
```diff
@@ -14,880 +14,772 @@
 #=== PRINT DETAILS ===
 logger.info("Verbose mode: #{$verbose}")
-if $verbose == "true"
+if $verbose == 'true'
```
same here?

FaragElsayed2 replied:
Done

Commit:
Adding ruby linting checks, Fixing some linting issues in drc files

atorkmabrains commented:
@proppy Anything else?

atorkmabrains commented:
@proppy Could you please merge?

proppy commented:
@atorkmabrains yes, sorry I was just waiting for the CI to pass.

Thanks for all the hard work on this!
